doc/stream-telemetry/stream-telemetry-hld.md

```diff
@@ -0,0 +1,986 @@
+
+The `Stream Telemetry Orch` is a new object within the Orchagent. It has following primary duties:
+
+1. Maintain the TAM SAI objects according to the stream telemetry configuration in the config DB.
+2. Generate a unique template ID for each stream telemetry profile to ensure distinct identification and management.
+3. Register and activate streams on counter syncd.
+
+`Stream Telemetry Orch` leverages `tam_counter_subscription` objects to bind monitoring objects, such as ports, buffers, or queues, to streams. Therefore, this orch must ensure that the lifecycle of `tam_counter_subscription` objects is within the lifecycle of their respective monitoring objects.
+
+#### Netlink Module and DMA Engine
+
+These two modules need to be provided by vendors. This document proposes a ring buffer communication model to support all expected TAM configurations as follows.
+
+![netlink_dma_channel](netlink_dma_channel.drawio.svg)
+
+### Data format
+
+We will use IPFIX as the report format, with all numbers in the IPFIX message in network-order (Big-endian).
+
+For more information on IPFIX, refer to the following resources:
+
+- [Specification of the IP Flow Information Export (IPFIX) Protocol for the Exchange of Flow Information](https://datatracker.ietf.org/doc/html/rfc7011)
+- [IP Flow Information Export (IPFIX) Entities](https://www.iana.org/assignments/ipfix/ipfix.xhtml)
+
+#### IPFIX header
+
+The `Version` and `Observation Domain ID` fields of the IPFIX header are identical for each IPFIX message.
+
+``` mermaid
+
+---
+title: stream message IPFIX header
+---
+packet-beta
+0-15: "Version = 0x000a"
+16-31: "Message Length = (16 + payload) bytes"
+32-63: "Export Timestamp: Second"
+64-95: "Sequence Number = 0, start from 0 and incremental sequence counter modulo 2^32"
+96-127: "Observation Domain ID = 0, always 0"
+
+```
+
+#### IPFIX template
+
+``` mermaid
+
+---
+title: stream message of IPFIX template
+---
+packet-beta
+0-15: "Set ID = 2"
+16-31: "Set Length = (12 + Number of Stats * 8) bytes"
+32-47: "Template ID = > 256 configured"
+48-63: "Number of Fields = 1 + Number of stats"
+64-79: "Element ID=observationTimeNanoseconds (325)"
+80-95: "Field length = 4 bytes"
+96-96: "1"
+97-111: "Element ID = Object index for the stats 1"
+112-127: "Field Length = 8 bytes"
+128-159: "Enterprise Number = SAI TYPE ID + SAI STATS ID for the stats 1"
+160-191: "..."
+192-192: "1"
+193-207: "Element ID = Object index for the stats N"
+208-223: "Field Length = 8 bytes"
+224-255: "Enterprise Number = SAI TYPE ID + SAI STATS ID for the stats N"
+
+```
+
+- For high-frequency counters, the native IPFIX timestamp unit of seconds is insufficient. Therefore, we introduce an additional element, `observationTimeNanoseconds`, for each record to meet our requirements.
+- The enterprise bit is always set to 1 for stats records.
+- The element ID of IPFIX is derived from the object index. For example, for `Ethernet5`, the element ID will be `0x5 | 0x8000 = 0x8005`.
```

Comment:
Having a variable element ID per SAI object will blow up the number of templates in the report (separate template ID per port, queue etc.).

I suggest having a fixed element ID for each stat, and introduce a new element `objectID`, similar to `observationTimeNanoseconds`, to put the object ID of a port or a queue in a data set itself.

Reply:
In my mind, I assume all stats(per port, queue etc) in one profile will be encoded into ONE template. I will generate a unique template ID for a profile.